lilio/calendar.py

```diff
@@ -27,27 +21,173 @@
+                are the same as the length keyword argument. Defaults to "0d".
+        """
+        self.length = length
+        self.gap = gap
+        self._role = role
+        self._target = role == "target"
+
+        self._gap_dateoffset: pd.DateOffset
+        self._length_dateoffset: pd.DateOffset
+
+        # TO DO: support lead_time
+        # self.lead_time = lead_time
+
+    @property
+    def is_target(self):
+        return self._target
+
+    @property
+    def role(self):
+        "Returns the type of interval."
+        return self._role
+
+    @property
+    def length(self):
+        "Returns the length of the interval, as a pandas.DateOffset."
+        return self._length
+
+    @length.setter
+    def length(self, value: Union[str, dict]):
+        self._length = value
+        if isinstance(value, str):
+            self._length_dateoffset = DateOffset(
+                **utils.parse_freqstr_to_dateoffset(value)
+            )
+        else:
+            self._length_dateoffset = DateOffset(**value)
+
+    @property
+    def length_dateoffset(self):
+        return self._length_dateoffset
+
+    @property
+    def gap(self):
+        """Returns the gap of the interval, as a pandas.DateOffset."""
+        return self._gap
+
+    @gap.setter
+    def gap(self, value: Union[str, dict]):
+        self._gap = value
+        if isinstance(value, str):
+            self._gap_dateoffset = DateOffset(
+                **utils.parse_freqstr_to_dateoffset(value)
+            )
+        else:
+            self._gap_dateoffset = DateOffset(**value)
+
+    @property
+    def gap_dateoffset(self):
+        return self._gap_dateoffset
+
+    def __repr__(self):
+        """String representation of the Interval class."""
+        props = [
+            ("role", self.role),
+            ("length", self.length),
+            ("gap", self.gap),
+        ]
+
+        propstr = ", ".join([f"{k}={repr(v)}" for k, v in props])
+        return f"{self.__class__.__name__}({propstr})"
+
+
+class Calendar:
+    """Build a calendar from scratch with basic construction elements."""
 
-    @abstractmethod
     def __init__(
         self,
         anchor: str,
-    ) -> None:
-        """For initializing calendars, the following five variables will be required."""
+        allow_overlap: bool = False,
+        mapping: Union[
+            None,
+            MappingYears,
+            MappingData,
+        ] = None,
+        intervals: Union[None, List[Interval]] = None,
+    ):
+        """Instantiate a basic container for building calendar using basic blocks.
+
+        This is a highly flexible calendar which allows the user to build their own
+        calendar with the basic building blocks of target and precursor periods.
+
+        Users have the freedom to create calendar with customized intervals, gap
+        between intervals, and even overlapped intervals. They need to manage the
+        calendar themselves.
+
```

Comment:
Shall we add a note here pointing to the shorthand calendar constructor functions?